CMakeLists.txt

```diff
@@ -405,9 +405,9 @@ set(SWIFT_STDLIB_MSVC_RUNTIME_LIBRARY
 
 
 if(BRIDGING_MODE STREQUAL "DEFAULT" OR NOT BRIDGING_MODE)
-  if(CMAKE_BUILD_TYPE STREQUAL "Debug" OR "${CMAKE_SYSTEM_NAME}" STREQUAL "Windows" OR (CMAKE_Swift_COMPILER AND CMAKE_Swift_COMPILER_VERSION VERSION_LESS 5.8))
+  if(CMAKE_BUILD_TYPE STREQUAL "Debug" OR "${SWIFT_HOST_VARIANT_SDK}" MATCHES "WINDOWS|ANDROID" OR (CMAKE_Swift_COMPILER AND CMAKE_Swift_COMPILER_VERSION VERSION_LESS 5.8))
```

Comment:
`CMAKE_SYSTEM_NAME` is not set by `build-script` when cross-compiling, this repo uses `SWIFT_HOST_VARIANT_SDK` instead.

Reply:
Most of the time `CMAKE_SYSTEM_NAME` is set by CMake. Scripts only need to provide a value explicitly when cross-compiling (see https://cmake.org/cmake/help/latest/variable/CMAKE_SYSTEM_NAME.html). 

I do not think the change here is correct. Before it was checking if the compiler was being build for Windows, and after the changes it is checking if the machine building the compiler is Windows (or Android). This is changing behaviour for some cross-compiling scenarios.

There's at least one check for `CMAKE_SYSTEM_NAME = Android`. Isn't that something that can be used?

I will recommend recovering the previous line, and adding the match for `ANDROID` in a different clause to not modify the previously working code:

```
if(CMAKE_BUILD_TYPE STREQUAL "Debug" OR "${CMAKE_SYSTEM_NAME}" STREQUAL "Windows" OR "${SWIFT_HOST_VARIANT_SDK}" STREQUAL "ANDROID" OR (CMAKE_Swift_COMPILER AND CMAKE_Swift_COMPILER_VERSION VERSION_LESS 5.8))
```

Reply:
> Most of the time CMAKE_SYSTEM_NAME is set by CMake. Scripts only need to provide a value explicitly when cross-compiling

Yep, that's the point I'm making. Suppose you want to cross-compile the Swift compiler from linux to Windows: the current check using `CMAKE_SYSTEM_NAME` would not work, because we do not set `CMAKE_SYSTEM_NAME` anywhere in `build-script`.

> I do not think the change here is correct. Before it was checking if the compiler was being build for Windows, and after the changes it is checking if the machine building the compiler is Windows (or Android). This is changing behaviour for some cross-compiling scenarios.

No, that's not how this repo works. [This repo only sets `SWIFT_HOST_VARIANT_SDK`](https://github.com/apple/swift/blob/3da894d704538557b1190f9089cd4163f868da08/utils/build-script-impl#L529) and [uses it extensively to implement its own cross-compilation system](https://github.com/apple/swift/blob/main/cmake/modules/AddSwift.cmake), without using CMake's built-in cross-compilation support that you refer to.

`SWIFT_HOST_VARIANT_SDK` does not refer to "the machine building the compiler," but _replaces_ `CMAKE_SYSTEM_NAME` as the host the compiler should be built for, at least in this repo.

> There's at least one check for CMAKE_SYSTEM_NAME = Android. Isn't that something that can be used?

It can be used when natively compiling, because as you noted, the native host can be inferred by CMake then. However, the few places that use `CMAKE_SYSTEM_NAME` in this repo all break when cross-compiling, for the reasons given above.

Reply:
`build-script` is just the way the Darwin/Linux CI uses to build Swift, but Windows CI builds, in particular, do not use `build-script` (it is impossible). Other people use unified builds with LLVM and other ways of kick starting the build and try to avoid `build-script`.

I would strongly recommend not changing anything related to other systems like Windows, specially when it is really easy to avoid it, like in this case.

Reply:
> build-script is just the way the Darwin/Linux CI uses to build Swift, but Windows CI builds, in particular, do not use build-script (it is impossible). Other people use unified builds with LLVM and other ways of kick starting the build and try to avoid build-script.

I'm well aware, as I have merged changes to those Windows batch files in this repo before.

> I would strongly recommend not changing anything related to other systems like Windows, specially when it is really easy to avoid it, like in this case.

Trust me, I know what I'm doing in this repo. Even those other build setups that [use `CMAKE_SYSTEM_NAME` have to translate that to the right `SWIFT_HOST_VARIANT_SDK` for this repo](https://github.com/apple/swift/blob/e5442fea506c74e1a1624566e25d73fa8115b69a/CMakeLists.txt#L142), or this repo's build does not work.

I had checked to make sure that translation was done above this line before changing this. I know what I'm doing by now. 😉

Reply:
Can I ask you please to not change this?

You don't know all "other build setups". There's a possibility of overriding `SWIFT_HOST_VARIANT_SDK` just some lines above: https://github.com/apple/swift/blob/e5442fea506c74e1a1624566e25d73fa8115b69a/CMakeLists.txt#L130-L132 . One can provide different values to `CMAKE_SYSTEM_NAME` and `SWIFT_HOST_VARIANT_SDK`.

Reply:
> Can I ask you please to not change this?

No, I'm confident in how this compiler build works. If you are not, please take a look at my links to the source to see what I'm talking about.

> One can provide different values to CMAKE_SYSTEM_NAME and SWIFT_HOST_VARIANT_SDK.

Sure, and then this compiler build will not work.

The reason that "override" you mention is there is because `build-script` passes in `SWIFT_HOST_VARIANT_SDK` to this CMake build, as I linked above. It _does not_ use `CMAKE_SYSTEM_NAME`. For those who insist on using `CMAKE_SYSTEM_NAME` instead outside of `build-script`, the translation to `SWIFT_HOST_VARIANT_SDK` that I linked you above is used.

Now, I can see how the use of both `CMAKE_SYSTEM_NAME` and `SWIFT_HOST_VARIANT_SDK` in this repo can be confusing, as I remarked that _both_ were being used in this repo from my first pull almost five years ago, though most heavy usage is of `SWIFT_HOST_VARIANT_SDK` alone.

That inconsistent usage happens to work when natively compiling because CMake can easily infer the native host and set `CMAKE_SYSTEM_NAME`, but it can break when cross-compiling if you don't set both variables correctly.

Now, let's look at the two scenarios, with `build-script` or without. `build-script` does not set a `CMAKE_SYSTEM_NAME` when cross-compiling, so it is safer to check `SWIFT_HOST_VARIANT_SDK` on this line then. Without `build-script`, one can do all kinds of nonsense, as you noted, but if you set `CMAKE_SYSTEM_NAME` alone, the translation above ensures that `SWIFT_HOST_VARIANT_SDK` is properly defined, so again it is safer to check `SWIFT_HOST_VARIANT_SDK` here.

I have worked through this logic multiple times, and I'm telling you, this is the right check for this repo.

Reply:
Not going to insist more on this topic. I am sorry we cannot agree in a solution that works for both cases. If nobody else have problems with this and I am the only one having this problem, feel free to merge.

Reply:
@drodriguez, I'm telling you, _this_ does work for both cases. If you just look elsewhere in the same file, there are [other `SWIFT_HOST_VARIANT_SDK` checks just like this](https://github.com/apple/swift/blob/e5442fea506c74e1a1624566e25d73fa8115b69a/CMakeLists.txt#L867), indicating at least that both styles are used.

If you are worried about breaking something, please just run the CI now, which as we discussed, runs both with and without `build-script`. That will show that this doesn't break anything.

Reply:
I think that @drodriguez has a valid point here.  Consider the builds for the macOS and the fat runtimes on Apple platforms.  These can be different (at least currently).  I think that we should prefer the CMake handling for the cross-compilation mechanism and is something that might be a possibility in the future.  We shouldn't make this harder for such a future refactoring.

Reply:
> Consider the builds for the macOS and the fat runtimes on Apple platforms. These can be different (at least currently).

[Every Darwin OS sets `SWIFT_HOST_VARIANT_SDK`](https://github.com/apple/swift/blob/71bbe17dc2dea9796d507b464846972d89a868d5/utils/build-script-impl#L592) as I'm checking here- see the next 100 lines after that link too- none set `CMAKE_SYSTEM_NAME`.

> I think that we should prefer the CMake handling for the cross-compilation mechanism

I know: you, me, and Daniel have had this argument since my first pull building the compiler for Android almost five years ago. The fact remains that after all these years, most of this main compiler/stdlib repo primarily uses `SWIFT_HOST_VARIANT_SDK` and with good reason, as we discussed before and here.

> and is something that might be a possibility in the future.

It hasn't happened in five years, I don't think it will happen anytime soon.

> We shouldn't make this harder for such a future refactoring.

It's a simple search-and-replace, it doesn't make it hard at all. On the contrary, checking `CMAKE_SYSTEM_NAME` as you did here breaks cross-compilation for every platform that doesn't set `CMAKE_SYSTEM_NAME` differently when cross-compiling, which is every platform using `build-script`, ie almost every Swift platform.

As I emphasized from my first pull, you can't check either `CMAKE_SYSTEM_NAME` and `SWIFT_HOST_VARIANT_SDK` randomly throughout this compiler codebase, because it will happen to work when natively compiling the compiler but breaks cross-compiling the Swift compiler and stdlib. Sneaking in `CMAKE_SYSTEM_NAME` checks like this piecemeal actually break things, it is not helping matters.

Since I'm probably the only person consistly cross-compiling the Swift compiler from one OS to another (CMake simply uses `Darwin`for all Apple platforms, so that wouldn't be considered cross-OS by CMake, I guess), I'm well aware of the problems with `CMAKE_SYSTEM_NAME`, which is why I removed its use for this check.

Now, all that said, I said from my first pull that I don't care whether `CMAKE_SYSTEM_NAME` or `SWIFT_HOST_VARIANT_SDK` is used in this repo, but one has to be chosen and used consistently and exclusively. Since this repo is currently a muddled mess, using `SWIFT_HOST_VARIANT_SDK` heavily but also using `CMAKE_SYSTEM_NAME` for a minority of checks, I'm switching this to the more well-supported `SWIFT_HOST_VARIANT_SDK`, which is the only one supported by `build-script`, while as I detailed above, it also works for platforms like Windows that only set `CMAKE_SYSTEM_NAME`, as the Windows CI run for this pull passing demonstrated.